dali/pipeline/data/tensor_vector.h

```diff
@@ -410,20 +667,58 @@ class DLL_PUBLIC TensorVector {
    */
   void UpdatePropertiesFromSamples(bool contiguous);
 
-  bool has_data() const;
-
   void resize_tensors(int size);
 
-  void update_view(int idx);
+  void recreate_views();
+
+  /**
+   * @brief Check if the metadata provided for new sample match the ones currently set for the batch
+   *
+   * When setting new sample, the source shape doesn't matter as it is adjusted for individual
+   * sample.
+   *
+   * When setting new sample the `shape_` must be adjusted.
+   *
+   * @param error_suffix Additional description added to the error message
+   */
+  void VerifySampleShareConformance(DALIDataType type, int sample_dim, TensorLayout layout,
+                                    bool pinned, AccessOrder order, int device_id,
+                                    const std::string &error_suffix = ".");
 
-  std::vector<std::shared_ptr<Tensor<Backend>>> tensors_;
+  /**
+   * @brief Check if the metadata provided for new sample match the ones currently set for the batch
+   *
+   * When copying new sample, pinned status and order of source and destination buffer can be
+   * different. Necessary synchronization is handled by the copy itself.
+   *
+   * When copying new sample the `shape_` must be adjusted.
+   *
+   * @param error_suffix Additional description added to the error message
+   */
+  void VerifySampleCopyConformance(DALIDataType type, int sample_dim, TensorLayout layout,
```

Comment:
Bikesheding bit more, `VerifySampleCopyCompability`?

Reply:
Why not

Reply:
Compatibility btw.